src/QsCompiler/Transformations/CallGraph.cs

```diff
@@ -194,20 +194,91 @@ public override int GetHashCode()
     /// </summary>
     public class CallGraph
     {
-        // TODO:
-        // This is the method that should be invoked to verify cycles of interest,
-        // i.e. where each callable in the cycle is type parametrized.
-        // It should probably generate diagnostics; I'll add the doc comment once its use is fully defined.
-        internal static bool VerifyCycle(CallGraphNode rootNode, params CallGraphEdge[] edges)
+        /// <summary>
+        /// Given a cycle of call graph edges, determines if the cycle is valid.
+        /// Invalid cycles are those that cause type parameters to be mapped to
+        /// other type parameters of the same callable (constricting resolutions)
+        /// or to a type containing a nested reference to the same type parameter,
+        /// i.e Foo.A -> Foo.A[].
+        /// Returns true if the cycle is valid, false if invalid.
+        /// </summary>
+        internal static bool VerifyCycle(params CallGraphEdge[] edges)
         {
-            var parent = rootNode.CallableName;
             var combination = new TypeResolutionCombination(edges.Select(edge => edge.ParamResolutions).ToArray());
-            var resolvedToConcrete = combination.CombinedResolutionDictionary.FilterByOrigin(parent)
-                .All(kvp => !(kvp.Value.Resolution is ResolvedTypeKind.TypeParameter tp) ||
-                    (tp.Item.Origin.Equals(kvp.Key.Item1) && tp.Item.TypeName.Equals(kvp.Key.Item2)));
-            return combination.IsValid && resolvedToConcrete;
-            //var isClosedCycle = validCycle && combined.Values.Any(res => res.Resolution is ResolvedTypeKind.TypeParameter tp && EqualsParent(tp.Item.Origin));
-            // TODO: check that monomorphization correctly processes closed cycles - meaning add a test...
+            if (!combination.IsValid)
+            {
+                return false;
+            }
+            var allResolutionsValid = combination.CombinedResolutionDictionary
+                .All(kvp => CheckTypeParameterResolutions.IsValidParamResolution(kvp.Key, kvp.Value));
+            return allResolutionsValid;
```

Comment:
`combination.IsValid` is true but `allResolutionsValid` could be false - it seems confusing that the combination could be valid but the resolutions inside the combination could be invalid?

Reply:
The first check, `combination.IsValid` comes from a tool that is agnostic to whether it is processing a cycle or a series of resolutions from a TypedExpression. Because of this it can't tell if the mapping `Foo.A -> Foo.A[]` is valid because that may be valid in a TypedExpression, if, for example the type parameter on the left is referring to a recursive reference while the type parameter on the left is referring to the calling context's type parameter. The second check is only valid in the case of cycle checking. In this case we know that `Foo.A -> Foo.A[]` is invalid, so we look for if that has occurred. Perhaps there is a better name for the variable for the second check? Or maybe we shouldn't capture it in a variable at all and just return the result of the expression, adding a comment to clarify what's being checked? What do you think?

Reply:
Hmm. I understand this is not how type variables are represented currently, but it seems like the mapping `Foo.A -> Foo.A[]` should always be invalid, and if those `A`s are different, they should have different (maybe auto-generated?) names.

> Perhaps there is a better name for the variable for the second check?

My issue is not only with the variable name `allResolutionsValid`, but with the method name `IsValidParamResolution`.

Reply:
Yes I agree, the type parameters need to have some way of differentiating between different instances of the same callable, but for now this is what we are stuck with. When that happens than `Foo.A -> Foo.A[]` where we are referring to the same type parameter would indeed always be not valid.

Yes I think I will remove the variable altogether and put a comment there, as well as rename the transformation and its static method.

Reply:
I've renamed the `IsValidParamResolution` method to `ContainsNestedSelfResolution`.

Reply:
Yes, exactly; TypeResolutionCombination should not allow this - if it does, I'd consider that a bug.